File: 준영/LV2/쿼드압축 후 개수 세기.py
def quard(arr, idx, temp, x, y):
    result = True
    zero_one = arr[x][y]
    for i in range(x, x+idx):
        for j in range(y, y+idx):
            if arr[i][j] != zero_one:
                result = False
    
    if result:
        for i in range(x, x+idx):
            for j in range(y, y+idx):
                temp[i][j] = 1

        return zero_one
    else:
        return -1

def solution(arr):
    n = len(arr)
    temp = [[0 for _ in range(n)] for _ in range(n)]
    answer = [0, 0]

    idx = n
    while idx >= 1:
        for i in range(0, n, idx):
            for j in range(0, n, idx):
                if temp[i][j] == 0:
                    zo = quard(arr, idx, temp, i, j)
                    if zo != -1:
                        answer[zo] += 1

        idx //= 2

    return answer


arr = [[1,1,0,0],[1,0,0,0],[1,0,0,1],[1,1,1,1]]	
solution(arr)

File: 준영/LV2/쿼드압축 후 개수 세기.js
function quard(arr, idx, visit, x, y) {
  var result = true
  var zero_one = arr[x][y]
  for (let i = x; i < x+idx; i++) {
    for (let j = y; j < y+idx; j++) {
     if (arr[i][j] != zero_one) {
       result = false
     } 
    }
  }
  
  if (result) {
    for (let i = x; i < x+idx; i++) {
      for (let j = y; j < y+idx; j++) {
        visit[i][j] = 1
      }
    }
    return zero_one
  } else {
    return -1
  }
  

}

function solution(arr) {
  var answer = [0, 0]
  var n = arr.length
  var visit = Array.from(Array(n), () => Array(n).fill(0))
  
  var idx = n

  while (idx >= 1) {
    for (let i = 0; i < n; i += idx) {
      for (let j = 0; j < n; j += idx) {
        if (visit[i][j] == 0) {
          var zo = quard(arr, idx, visit, i, j)
          if (zo != -1) answer[zo] += 1
        }
      }
    }
    idx = parseInt(idx/2)
  }
  return answer
}


arr = [[1,1,0,0],[1,0,0,0],[1,0,0,1],[1,1,1,1]]	
solution(arr)